// components/Draft/RemoveCard/index.js
import React, {useState} from 'react'

import classes from './style.module.css'

// Вот этот компонент мне нравится больше всего, ам ням

export class  RemoveCard extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            id: '',
            text: ''
        }

        this.addInput = this.addInput.bind(this)
    }

    addInput(noteId) {
        this.setState({
            id: noteId
        })
    }

    checkForm(e, id) {
        e.preventDefault();
       
        if (this.state.text.trim()) {
            this.props.renameDraft(this.state.text , id)
            this.setState({
                text: ''
            })
            e.target.firstChild.value = '';
        } else {
            return null
        }
    }

    // Не ругайтесь за запутанный jsx это я эксперементировал с написанием

    render() {
        return (
            <div>
                {this.props.notes.map(note => {
                    return (
                        <div className = {classes.card} key = {note.id} >
                            <div className = {classes.text} 
                                onClick = {() => this.addInput(note.id)}
                            >   
                                {note.text}
                                <form 
                                    onSubmit = {e => this.checkForm(e, note.id)}
                                    className = {(this.state.id === note.id) ? 
                                    classes.visible : classes.unvisible}
                                >
                                    <input onChange={e => this.state.text = e.target.value}></input>
                                    <button>Изменить</button>
                                </form>
                            </div>
                            <button 
                                className = {classes.remove} 
                                onClick = {() => this.props.onRemove(note.id)}
                            >
                                Remove
                            </button>
                            <button className = {classes.save} onClick = {() => {
                                this.props.onRemove(note.id);
                                this.props.addInSave(note.text, "save")
                            }}>
                                Save
                            </button>
                        </div>
                    )
                })}
            </div>
        )
    }
}
